Workflows, automation, esbuild

ktarmyshov opened the pull request:
Hi @geoffrich!

I have invested some time to update the dependencies & workflows for better automation. This was for me as an intermediate step towards setting up the same for my own app.

Please check and tell me if you want to have this merged. It requires some setup in the repo (some bots, secrets, etc.) to work. If you want to have it merged, I will prepare a detailed setup or we can have a call to do it together, which probably will be much faster.

Here is the summary:
- For the workflows I used what was done in @sveltejs/*
- Migrate release flow from google release-please to changesets (https://github.com/changesets/changesets)
  - In general the flow is the same, the difference, is that they use MD files to capture the changes, so squash merge is possible
  - They also have the bot to add and update comments on the incoming pull requests, to the requesting user can already describe the changes in detail
- Re-work CI workflow (again @sveltejs is the reference here)
  - Added testing node matrix 18 & 20
  - Added testing on the SWA - here the prerender option is disabled until https://github.com/Azure/static-web-apps-cli/issues/945 is resolved
  - Rework testing on azure with concurrency - here are some issues with the github, as it allows only 1 pending job, others will be cancelled. This is unfortunate, but this should not be a very big deal, as the failed/cancelled jobs can be resumed, without restarting already completed. Also once the 20.x runtime is deployed, 18.x can be removed from the matrix, and this should pretty much remove the issue
  - Azure Testing is done only for internal pull requests
  - protection for main branch via rules can and should be enabled
- Dependabot: PRs are created, changeset is created, tested in swa & azure and if successful, and (only for minor and patch) PRs are automatically merged into the main (but release is always manual)

I think that's all.

Here are the examples how it is executed
- push on main https://github.com/ETF-Math/svelte-adapter-azure-swa/actions/runs/13822201463
- Internal pull request - this is for created release branch https://github.com/ETF-Math/svelte-adapter-azure-swa/pull/15
- Internal pull request merged
  - https://github.com/ETF-Math/svelte-adapter-azure-swa/actions/runs/13821314755
  - https://github.com/ETF-Math/svelte-adapter-azure-swa/pull/14
- External pull request
  - https://github.com/ETF-Math/svelte-adapter-azure-swa/pull/9
- External PRs do not run the tests in azure, but the job is executed (without any actions) to deliver the status to the merge protection

Cheers,
Konstantin

ktarmyshov commented:
Need to rearrange some things, will PR later

geoffrich commented:
Hey, thanks for reaching out. Before you get too far into this:

- Right now I'm not interested in migrating to changesets. The current release process is fine for my purposes, so I'd rather not spend effort switching over. This repo is also only a single package so changesets could be overkill
- Dependabot I'm meh on - this adapter has very few production dependencies, so I'm not sure I want to spend effort getting auto-dependency upgrades set up
- Testing improvements I'm open to, but any large changes should have an issue opened first to discuss what the improvements would be

For future work, please take a look at the [contributing guidelines](https://github.com/geoffrich/svelte-adapter-azure-swa/blob/main/CONTRIBUTING.md#how-to-contribute) -- any major changes like this should have an issue opened first so we can discuss whether the change makes sense for the repo before you take time implementing it.

Huge PRs like this are also very difficult to review. Ideally, each major change should have its own PR, so that it's easy to understand what changes correspond to which features. For example, this PR seems to include multiple changes that could be their own PR:
- migrate to changesets
- enable dependabot
- update testing workflows

Small PRs are easier to review than large ones, and the harder a PR is to review the longer it takes to get merged.

I appreciate your desire to contribute to this repo, just want to make the process smoother for both of us!

I also want to note that I'm about to go on leave for a few weeks, so I may be slow to respond or review changes.

ktarmyshov commented:
Hi, @geoffrich!

Have a nice vacation.

I gave it a second thought. Most of the stuff in the PR is a question of taste. So I would not create a PR for this. As I said before, I was experimenting with the flows and change tracking as an intermediate step for my repo.

So if you are interested, I would suggest to take a look at the [fork](https://github.com/ktarmyshov/svelte-adapter-azure-swa/tree/main), and see what you would want to take over. Here is a short summary:
- Workflows
  - CI
    - added test on SWA (modified demo app to react to SWA env and switch off the prerender until MS swa cli issue is resolved)
    - adjusted the flow, so branch protection can be used
  - Dependabot - automated test and merge
  - Release - via changeset, trigger on success of CI, added provenance
- Demo app - adjusted for SWA test

Konstantin
